lib/DiscoveryStreamFeed.jsm

```diff
@@ -1033,8 +1052,18 @@ this.DiscoveryStreamFeed = class DiscoveryStreamFeed {
               data: this.config,
             }));
             break;
+          case PREF_TOPSTORIES:
+            if (!action.data.value) {
+              // Ensure we delete any remote data potentially related to spocs.
+              this.clearSpocs();
+            }
+            break;
           // Check if spocs was disabled. Remove them if they were.
           case PREF_SHOW_SPONSORED:
+            if (!action.data.value) {
+              // Ensure we delete any remote data potentially related to spocs.
+              this.clearSpocs();
+            }
             await this.loadSpocs(update => this.store.dispatch(ac.BroadcastToContent(update)));
```

Comment:
Should we be doing this for both spocs pref turning on and off? And this is making me wonder if we should be doing this as well for `PREF_TOPSTORIES` (i.e., have the two `case` statements use the same body ??)

Reply:
It actually already works this way I think.

I asked some of these questions this morning too.

From what I can tell, loadSpocs with spocs turned off successfully loads no spocs, so if you went from on to off without calling loadSpocs again, it would require a refresh of the page or restart to remove spocs.

On top of that, changing PREF_TOPSTORIES fires a chain of events that also ends up ref retching spocs because it clears cache and fetches fresh if that pref changes.

Reply:
If this is not actually the case though, I do think what you expected to happen should one way or another.

Reply:
I don't see `loadSpocs` getting called when enabling/disabling topstories pref without this PR. The original comment "Remove them if they were [enabled]" makes it seem like it needs to clean up stuff, but looks like the removing is for tab content and not necessarily cached state. I'm not sure if we need to clean up the disk cache in addition to cleaning up server data. Will the local cached data have personalized data based on the server response going forwards?

Reply:
Hm loadSpocs not being called, it should go into here: https://github.com/mozilla/activity-stream/blob/master/lib/DiscoveryStreamFeed.jsm#L879

Which should reset everything, and then re enable everything, which calls `enable`, then `refreshAll`, then `loadSpocs`